x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Expression.java

```diff
@@ -207,4 +223,34 @@ public String toString() {
     public String propertiesToString(boolean skipIfChild) {
         return super.propertiesToString(false);
     }
+
+    protected static boolean isAttributePushable(Expression expression, boolean supportMetadataFields, FieldInfo fieldInfo) {
+        if (isPushableFieldAttribute(expression, fieldInfo)) {
+            return true;
+        }
+        if (expression instanceof MetadataAttribute ma && ma.searchable()) {
+            return supportMetadataFields;
+        }
+        return false;
+    }
+
+    private static boolean isPushableFieldAttribute(Expression exp, FieldInfo fieldInfo) {
+        if (exp instanceof FieldAttribute fa && fa.getExactInfo().hasExact() && isAggregatable(fa)) {
+            return fa.dataType() != DataType.TEXT || fieldInfo.hasIdenticalDelegate(fa);
+        }
+        return false;
+    }
+
+    /**
+     * this method is supposed to be used to define if a field can be used for exact push down (eg. sort or filter).
+     * "aggregatable" is the most accurate information we can have from field_caps as of now.
+     * Pushing down operations on fields that are not aggregatable would result in an error.
+     */
+    protected static boolean isAggregatable(FieldAttribute f) {
+        return f.exactAttribute().field().isAggregatable();
+    }
+
+    public interface FieldInfo {
+        boolean hasIdenticalDelegate(FieldAttribute attr);
```

Comment:
This should probably have some javadoc for what this predicate checks.  Just copying the bit from `PlannerUtils` that's getting deleted would be enough, I think.

Reply:
good idea